zanata-war/src/main/java/org/zanata/security/ZanataIdentity.java

```diff
@@ -182,7 +185,27 @@ public boolean hasPermission(String action, Object... targets) {
      *             if logged in but not authorised
      */
     public void checkPermission(String action, Object... targets) {
-        super.checkPermission(MultiTargetList.fromTargets(targets), action);
+        try {
+            super.checkPermission(MultiTargetList.fromTargets(targets), action);
+        } catch (AuthorizationException exception) {
+            // try to produce a better than default error message
+            List<String> meaningfulTargets = Lists.newArrayList();
+            for (Object target : targets) {
+                if (target instanceof HasUserFriendlyToString) {
+                    String targetString = ((HasUserFriendlyToString) target).userFriendlyToString();
+                    meaningfulTargets.add(targetString);
+                } else {
+                    log.warn(
+                            "target [{}] may not have user friendly string representation",
+                            target.getClass());
+                    meaningfulTargets.add(target.toString());
+                }
+            }
+            throw new AuthorizationException(
+                    String.format(
+                            "Failed to obtain permission('%s') with following facts(%s)",
+                            action, meaningfulTargets));
```

Comment:
Is `facts` technical jargon here about the permission checking? For users it might be better to say "information", it is probably fine though as long as there is enough in the detail.